### ISSUE TITLE
Autoreload slow

### ISSUE THREAD

**Issue by andyljones** [ipython/ipython#11817]
On upgrading from 7.5 to 7.6.1, autoreload in Jupyter Notebook is much slower, to the point of seeming to hang for a >minute. When I lose my patience and interrupt execution, I get a traceback like this

```
[autoreload of algos.ppo.nets failed: Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/IPython/extensions/autoreload.py", line 245, in check
    superreload(m, reload, self.old_objects)
  File "/opt/conda/lib/python3.7/site-packages/IPython/extensions/autoreload.py", line 450, in superreload
    update_generic(old_obj, new_obj)
  File "/opt/conda/lib/python3.7/site-packages/IPython/extensions/autoreload.py", line 387, in update_generic
    update(a, b)
  File "/opt/conda/lib/python3.7/site-packages/IPython/extensions/autoreload.py", line 357, in update_class
    update_instances(old, new)
  File "/opt/conda/lib/python3.7/site-packages/IPython/extensions/autoreload.py", line 317, in update_instances
    update_instances(old, new, obj, visited)
  File "/opt/conda/lib/python3.7/site-packages/IPython/extensions/autoreload.py", line 317, in update_instances
    update_instances(old, new, obj, visited)
  File "/opt/conda/lib/python3.7/site-packages/IPython/extensions/autoreload.py", line 317, in update_instances
    update_instances(old, new, obj, visited)
  [Previous line repeated 2 more times]
  File "/opt/conda/lib/python3.7/site-packages/IPython/extensions/autoreload.py", line 300, in update_instances
    for obj in (obj for obj in objects if id(obj) not in visited):
KeyboardInterrupt
]
```

I don't have a minimal example I'm afraid, but if anyone arrives here Googling for a similar problem, rolling back to 7.5 fixes it.

My specs are

Ubuntu 18.04.2 LTS
Python 3.73
Jupyter:
```
jupyter==1.0.0
jupyter-client==5.3.0
jupyter-console==6.0.0
jupyter-contrib-core==0.3.3
jupyter-contrib-nbextensions==0.5.1
jupyter-core==4.5.0
jupyter-highlight-selected-word==0.2.0
jupyter-latex-envs==1.4.6
jupyter-nbextensions-configurator==0.4.1
```

**Comment by Carreau** [comment]
Hum, thanks for the report; i'll see if I can reproduce it; or at least add some debugs informations to a branch, and have you tried tha branch to figure out what's happening.

**Comment by smerzbach** [comment]
I can't add further debug info, but I had a similar problem with ipython-7.6.1 and Spyder-3.3.6:
 * open ipython console
 * select some code in the editor
 * hit F9 (this essentially copies the code to the console and executes it)

Every other time, this causes ipython to hang until interrupted by ctr+c, which then gives a stacktrace as pasted above. Repeating the exact same procedure works fine on the next try.
Downgrading to ipython-7.5 resolved the issues.

This is all on a Ubuntu 16.04, with Spyder and ipython installed via pip3:

    pip3 install --user --upgrade spyder3

To fix the issue:

    pip3 install --user ipython==7.5

Is there further info I could provide?

**Comment by WestXu** [comment]
Same problem. 

Add a few extra info:

1. The larger the RAM cost of the session is, the longer the hanging lasts.
2. If you wait until the hanging finishes, the reload **actually worked** fine. All instance of classes were reload as expected.
3. If you stop the hanging in the middle, again the reload **actually worked**, but the autoreload **won't work agian** from now on in this IPython session.
4. If you stop the hanging real fast like **right after** you execute the cell, the autoreload **won't work** this time and won't work in this IPython session.

**Comment by daharn** [comment]
This is most likely linked to the inclusion of class instance reloading in 7.6. It uses recursion to scan through the all user objects in the current session so it makes sense that it takes longer for larger session. Question is now if this is a bug in the sense that the scanning doesn't work as intended or if this kind of scanning just takes unacceptably long for certain session states.

A dump of a session where this is a problem might be helpful...

**Comment by bnaul** [comment]
This problem seems to be extra bad when I'm working with large pandas `DataFrame`s; is it possible that every single string value is getting scanned by the autoreload logic? If so is there any possible way to work around this case?

**Comment by daharn** [comment]
It might be worth considering to separate the instance-updating mechanism from the original autoreload functionality until this problem is further inspected. 

A quick-and-dirty solution (apart from downgrading) would be to comment out the line
https://github.com/ipython/ipython/blob/1879ed27bb0ec3be5fee499ac177ad14a9ef7cfd/IPython/extensions/autoreload.py#L357

**Comment by syagev** [comment]
> A dump of a session where this is a problem might be helpful...

@daharn I encounter this issue quite consistently when there are large pandas DataFrames around as @bnaul pointed out. How do I get you this "session dump"?

**Comment by daharn** [comment]
> > A dump of a session where this is a problem might be helpful...
> 
> @daharn I encounter this issue quite consistently when there are large pandas DataFrames around as @bnaul pointed out. How do I get you this "session dump"?

Although I believe I ran into this issue myself now for some time, if you have a session that reproduces this consistently, you could use pickle (or dill) to save it to a file and upload it here. I probably wont' be able look into this for the next couple of weeks, but maybe someone else is.

**Comment by syagev** [comment]
In the attached session, changing any of the imported modules causes autoreload to hang:
https://drive.google.com/file/d/1auE2ssFabG6Pgwl7jZPA2yRTqg6-xUcV/view?usp=sharing

Since I guess this dump contains much of my source code (due to nested imports and such) I'm going to have to share privately... so you need to use a google drive account and lmk when you request access so I'll immediately approve... same goes for anyone else who has time to look at it.

**Comment by daharn** [comment]
> In the attached session, changing any of the imported modules causes autoreload to hang:
> https://drive.google.com/file/d/1auE2ssFabG6Pgwl7jZPA2yRTqg6-xUcV/view?usp=sharing
> 
> Since I guess this dump contains much of my source code (due to nested imports and such) I'm going to have to share privately... so you need to use a google drive account and lmk when you request access so I'll immediately approve... same goes for anyone else who has time to look at it.

Sent you a request (Niclas)

**Comment by rsebbe** [comment]
I confirm the issue with 7.6.1 and 7.7.0. Hangs a couple minutes within `autoreload.py` after editing library code and re-executing a cell, when substantial amount of data has been previously loaded.

Workaround that works here - revert to ipython 7.5.0:
`conda install ipython=7.5.0`

**Comment by Carreau** [comment]
Maybe simply iterating of gc.get_objects() instead of what we are doing now
would be sufficient. Though someone need to find the time to write that
code and test it.

There might also be another way to get all object of a given class.

I unfortunately have limited time to do it myself.

**Comment by Carreau** [comment]
Would you be willing to try #11849 ? This may be faster; but this may be slower as well. 
I have a potential other solution is this is still too slow.

**Comment by cshimmin** [comment]
I am experiencing this issue as well; having a several gigabyte numpy array loaded in to memory, autoreload hangs for at least several minutes making it completely useless; I haven't had the patience to see if it ever finishes. The problem only just started when I upgraded several packages today; unfortunately I don't know what version I was using before.

I tried @daharn's suggestion to comment out `update_instances(new, old)` in autoreload.py, but this line is not present in my version.

```
jupyterlab==1.0.5
ipykernel==5.1.2
ipython==7.7.0
```

**Comment by daharn** [comment]
> I tried @daharn's suggestion to comment out `update_instances(new, old)` in autoreload.py, but this line is not present in my version.
> 
> ```
> jupyterlab==1.0.5
> ipykernel==5.1.2
> ipython==7.7.0
> ```

That is odd,  this line should be there from https://github.com/ipython/ipython/commit/bd5753b3b1d41dfaf79c694a7daaf1cd62a720f6 onwards. How did you install your version?

**Comment by cshimmin** [comment]
@daharn, my mistake! I accidentally opened up the version of `autoreload.py` from a different virtualenv. After commenting out the line in the correct file, I can confirm that this does appear to fix the issue with hanging. Many thanks!

Perhaps if not simple long term fix works, the class instance updating feature can be disabled by an optional argument to the `%autoreload` magic?

**Comment by silencial** [comment]
Same issue here with ipython=7.7.0, change to ipython=7.5.0 fix this.

**Comment by Carreau** [comment]
Locking the issue

Only commenting that you have the same issue is not constructive. 

If you want to help try the suggested fix  on #11849

**Comment by Carreau** [comment]
Should be closed by #11876
